contracts/ligo/src/swaps/README.md

```diff
@@ -0,0 +1,159 @@
+  }
+```
+
+A `swap_info` type represents all the full information stored about a single swap.
+
+```ocaml
+type swap_info =
+  [@layout:comb]
+  { swap_offer : swap_offer
+  ; seller : address
+  }
+```
+
+#### Storage
+
+Entries in storage include:
+* `%swaps : (swap_id ,swap_info) big_map` with all the swaps.
+
+### Entrypoints
+
+Notes:
+* By *transfer* (in the context of FA2 tokens) we mean "performing a zero-xtz transaction that invokes FA2 `transfer` entrypoint".
+* Entrypoints that perform FA2 transfer may throw the respective errors enlisted in [FA2 interface](https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-12/tzip-12.md#error-handling).
+
+#### %start
+
+```ocaml
+| Start of swap_offer
+```
+
+Proposes an exchange.
+
+This transfers the `assets_offered` tokens from the `sender` to our contract (thus assuming that our contract has been made an operator of the sender's tokens).
+
+The swap will have `Open` state, and its `seller` will be initialized to the current `sender`.
+
+The contract automatically assigns a `swap_id` (basing on an auto-incrementing counter) to the swap. We assume that it is the caller's responsibility to extract and record the `swap_id`, it will be available in the applied operation.
+
+In case if some of `assets_offered.fa2_address` refers to an invalid FA2 contract (not originated or without `transfer` entrypoint), `SWAP_OFFERED_FA2_INVALID` is raised.
+
+In case if tokens transfer from `sender` to our contract fails (due to lack of tokens or other issues), the respective FA2 error is propagated.
+
+`assets_requested` is not validated and must be checked by the client for a better UX.
+
+We do not check whether any assets list is empty on purpose.
+
+#### %cancel
+
+```ocaml
+| Cancel of swap_id
+```
+
+Marks the swap as canceled and returns the tokens to the `seller`. Further operations with this swap won't be possible.
+
+Can be called only by the `seller`, otherwise `NOT_SWAP_SELLER` error is raised.
+
+The swap gets removed from big_map, any further operation for this swap will raise `SWAP_NOT_EXIST`.
+
+#### %accept
+
+```ocaml
+| Accept of swap_id
+```
+
+This
+1. Transfers the offered assets from our contract to the `sender`;
+2. Transfers the requested assets from the `sender` to the `seller`.
+
+Marks the swap as finished by the current `sender`. Further operations with this swap won't be possible.
+
+This can be called by *any* address.
+
+The swap gets removed from big_map, any further operation for this swap will raise `SWAP_NOT_EXIST`.
+
+In case if some of `assets_requested.fa2_address` refers to an invalid FA2 contract (not originated or without `transfer` entrypoint), `SWAP_REQUESTED_FA2_INVALID` is raised. This error normally should not happen, as the client is supposed to validate `assets_requested` on swap start.
+
+In case if any tokens transfer mentioned above fails (due to lack of tokens or other issues), the respective FA2 error is propagated.
```

Comment:
```suggestion
In case any tokens transfer mentioned above fails (due to lack of tokens or other issues), the respective FA2 error is propagated.
```